---
title: "DNA Damage Estimation"
author: "Peter Diakumis"
date: "14 July 2017"
output: 
  html_document: 
    keep_md: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, comment = NA)
```
```{r message=FALSE}
library(ggplot2)
library(ggforce)
library(dplyr)
library(readr)
```

Here we'll run the DNA Damage Estimator scripts from the
[Ettwiller GitHub repo](https://github.com/Ettwiller/Damage-estimator).

# Example Run
An example BAM file was downloaded from
[bds](https://github.com/vsbuffalo/bds-files/tree/master/chapter-11-alignment):

```{r engine='bash'}
ls -lLh ../data/example
```

## Step 1: Split BAM into R1 and R2

### Command Line

```{r engine='bash', eval=F}
perl ../scripts/0-split_mapped_reads.pl \
  --bam ../data/example/NA12891_CEU_sample.bam \
  --genome ../data/genome/human_g1k_v37.fasta \
  --mpileup1 ../data/out/out1.mpileup \
  --mpileup2 ../data/out/out2.mpileup
```

### Output

```{r engine='bash'}
ls -lLh ../data/out
```

```{r engine='bash'}
head -n5 ../data/out/*.mpileup
```

The script uses `samtools mpileup` to output a summary of the read pileup in the
given BAM file. Options used are:

* `-O`: output base positions on reads 
* `-s`: output mapping quality
* `-q`: skip alignments with mapQ smaller than [10]
* `-Q`: skip bases with baseQ/BAQ smaller than [0]


## Step 2: Estimate basic damage 

### Command Line
```{r engine='bash', eval=F}
perl ../scripts/1a-estimate_damage.pl \
  --mpileup1 ../data/out/out1.mpileup \
  --mpileup2 ../data/out/out2.mpileup \
  --id foo \
  > ../data/out/foo.damage
```

### Output

```{r engine='bash'}
head -n5 ../data/out/foo.damage
```

Column description:

1. raw count of variant type
2. variant type (ex. G_T, G to T)
3. id (from the --id option)
4. frequency of variant
5. family (the variant type and reverse complement)
6. GIV-score

If you have followed the standard protocol for acoustic shearing during library preparation you should obtain a GIV score for G_T around 2.

### Plot

```{r example_plot1}
type_clean <-c("G_T", "C_A", "C_T", "G_A", "T_A", "A_T",
               "A_G", "T_C", "C_G", "G_C", "T_G", "A_C")
mut <- readr::read_tsv("../data/out/foo.damage",
                       col_names =  c("abs", "type", "experiment", "count", "family", "damage"),
                       col_types = "iccdcd") %>% 
  filter(type %in% type_clean) %>% 
  mutate(type = factor(type, level = type_clean))

#coloring scheme (feel free to change)
local_color <- c("cornflowerblue", "royalblue4", paste0("grey", c(1, seq(10, 100, 10))))

g <- ggplot(mut, aes(x = reorder(type, damage), y = log2(damage), color = experiment))

g + geom_point(alpha = 0.6, size=1.5) +
  scale_colour_manual(values = local_color) +
  geom_hline(yintercept = log2(1.5), color = "#990000", linetype = "dashed") +
  annotate("text", x = 4, y = log2(1.6), color = "#990000",
           label = "Above this line 1/3 of variants is due to damage") +
  geom_hline(yintercept = 0, color = "grey") +
  theme(panel.background = element_rect(fill = 'white', colour = 'white'), 
        legend.position = "bottom", axis.text.x = element_text(angle = 90, hjust = 1, size=11)) +
  ggtitle("GIV scores for variant types")
```


## Step 3: Estimate damage relative to read position

### Command Line
```{r engine='bash', eval=F}
perl ../scripts/2a-estimate_damage_location.pl \
  --mpileup1 ../data/out/out1.mpileup \
  --mpileup2 ../data/out/out2.mpileup \
  --id foo2 \
  --out ../data/out/foo2.damage
```

### Output

```{r engine='bash'}
head -n5 ../data/out/foo2.damage
```

Column description:

1. id (from the --id option)
2. variant type (ex. G_T, G to T)
3. R1 or R2
4. count (freq)
5. absolute counts
6. position on the read

### Plot

```{r example_plot2, fig.height=20}
mut <- readr::read_tsv("../data/out/foo2.damage",
                       col_names = c("experiment", "type", "read", "count", "abs", "loc"),
                       col_types = c("cccdii"))

ggplot(mut) +
  geom_point(aes(x = loc, y = count)) +
  theme_bw() +
  facet_grid(type~read, scales = "fixed")
```

