# DDE Method Summary {#dde-summary}

DDE has 5 main steps:

1. Split BAM into R1 and R2 BAMs
2. Generate R1 and R2 pileups
3. Clean bases in pileups
4. Get relative counts for pileups
5. Estimate final damage metrics based on relative counts from step 4.

## `split_mapped_reads.pl`

* Splits a `BAM` file into two BAM files containing the first mate (R1 - flag 64) and the
second mate (R2 - flag 128) of the paired-end reads, using `samtools view -f <flag>`.

* Creates `pileup` files using `samtools mpileup -O -s -q <min_mapq> -Q <min_baseq>`, where:
    - `-O`: output base positions on reads
    - `-s`: output mapping quality
    - `min_mapq` = 10 by default
    - `min_baseq` = 0 by default

### Pileup format

In the pileup format (without -u or -g), each line represents
a genomic position, consisting of:

| column   | description                |
|----------|----------------------------|
| chr      | chromosome name            |
| pos      | 1-based coordinate         |
| ref      | reference base             |
| cov      | depth of coverage at site  |
| rbase    | read bases                 |
| baseq    | base qualities             |

Information on match/mismatch, indel, strand, mapping quality and start/end of
a read are all encoded at the __read bases__ column. In this column:

- dot: match ref on forward strand
- comma:  match ref on reverse strand
- `>` or `<`: reference skip,
- `[ACGTN]`: mismatch on the forward strand
- `[acgtn]`: mismatch on the reverse strand.
- `\\+[0-9]+[ACGTNacgtn]+` pattern: insertion between this and next ref position
    - Integer = length of insertion, followed by inserted sequence.
- `-[0-9]+[ACGTNacgtn]+` pattern : deletion from the reference.
    - Deleted bases will be presented as `*` in the following lines.
- `^`: start of a read. The ASCII of the character following `^` minus 33 gives
  the mapping quality.
- `$`: end of a read.

#### Pileup Examples

* Row1: vanilla
* Row2: with `--no-BAQ`
* Row3: with `-O` (base positions on reads)
* Row4: with `-s` (mapping quality)

```
$> samtools mpileup --region 1:215906528-215906567 --fasta-ref human_g1k_v37.fasta.gz NA12891_CEU_sample.bam
1	215906528	G	21	,,,,,,,,.,,,.,,..,.,,	;=?./:?>>;=7?>>@A?==:
1	215906528	G	21	,,,,,,,,.,,,.,,..,.,,	;=?./:?>>;=7?>>@A?==:
1	215906528	G	21	,,,,,,,,.,,,.,,..,.,,	;=?./:?>>;=7?>>@A?==:	49,42,33,32,44,31,28,40,20,18,28,25,15,15,13,12,11,7,5,21,18
1	215906528	G	21	,,,,,,,,.,,,.,,..,.,,	;=?./:?>>;=7?>>@A?==:	]]]]F]]]F]FF]]]]:]]FF
[...]
1	215906534	G	19	,,,,.,,.,,..,.,,,,^].	=9?<>;;?>=@B>>??13>
1	215906534	G	19	,,,,.,,.,,..,.,,,,^].	=9?<>;;?>=@B>>??13>
1	215906534	G	19	,,,,.,,.,,..,.,,,,^].	=9?<>;;?>=@B>>??13>	48,37,34,46,26,24,34,21,21,19,18,17,13,11,27,24,21,2,1
1	215906534	G	19	,,,,.,,.,,..,.,,,,^].	=9?<>;;?>=@B>>??13>	]]]]F]F]]]]:]]FFF]]
[...]
1	215906539	C	14	,,$.,,,+1g,,....,.	6;244.76>15:6:
1	215906539	C	14	,,$.,,,+1g,,....,.	6;244.76>15:6:
1	215906539	C	14	,,$.,,,+1g,,....,.	6;244.76>15:6:	42,51,31,29,39,36,26,24,23,22,16,8,7,6
1	215906539	C	14	,,$.,,,+1g,,....,.	6;244.76>15:6:	]]F]FF]]]:]F]]
[...]
1	215906547	C	7	,,,....	086=C>=
1	215906547	C	15	gGg$,GggGG,,....	<;80;><9=86=C>=
1	215906547	C	7	,,,....	086=C>=	45,26,15,14,8,4,3
1	215906547	C	7	,,,....	086=C>=	F]]]F:S
[...]
1	215906555	G	12	aaaaAAAAaAAA	>?8?;:335?:A
1	215906555	G	16	.$aaaaaA.AAAaAAA^:A	2@>?8?;<:335?:A>
1	215906555	G	12	aaaaAAAAaAAA	>?8?;:335?:A	48,45,42,23,22,16,12,11,10,8,7,6
1	215906555	G	12	aaaaAAAAaAAA	>?8?;:335?:A	FFF]]F:SF]]:
[...]
```

So in summary, the `pileup` files generated from this script contain:

chr, pos, ref, cov, rbases, baseq, mapq, basepos:

```
[...]
1       215906546       G       6       ..,...  (!;;><  ]]]]:S  33,30,25,13,3,2
1       215906547       C       6       GG,...  !!8=>=  ]]]]:S  34,31,26,14,4,3
1       215906548       G       6       .C,...  !!=746  ]]]]:S  35,32,27,15,5,4
1       215906549       G       6       .$.,... !!>7?;  ]]]]:S  36,33,28,16,6,5
1       215906550       G       5       .,...   !>6?8   ]]]:S   34,29,17,7,6
1       215906551       G       5       .,...   !>7?5   ]]]:S   35,30,18,8,7
1       215906552       G       5       .$,...  !><87   ]]]:S   36,31,19,9,8
1       215906553       G       4       ,...    :<?9    ]]:S    32,20,10,9
1       215906554       C       4       ,...    @483    ]]:S    33,21,11,10
1       215906555       G       4       aAAA    $;31    ]]:S    34,22,12,11
[...]
```

## `estimate_damage.pl`

* Parameters:
    - pileups R1, R2
    - minimum base quality score (default 30)
    - minimum coverage (default 1)
    - maximum coverage (default 100)

Has two functions:

* Function1: `clean_bases`: called as `clean_bases(<rbases>, <cov>)`. The `<cov>` parameter
  isn't used anywhere in the function (sigh). The `<rbases>` is referred to as `pattern`.
    1. Take the `pattern` and remove any `^` followed by a character, and any `$`.
    2. If there is an indel in the `pattern`, then remove as many characters as
       its length, leaving the `+` or `-` sign.
    3. Return the `pattern`.

Here's a sample of what the function does:

```
dirty: ,.,,+1a,
clean: ,.,+,
dirty: .$.+1T,,..,.,.
clean: .+,,..,.,.
dirty: ,.,+1a,
clean: ,.+,
dirty: .+1T,$,,,,....,,.....
clean: +,,,,,....,,.....
dirty: ,,,.+1T.,,.....A.
clean: ,,,+.,,.....A.
dirty: ,.+1T,$.,.,,.,,..,
clean: ,+,.,.,,.,,..,
dirty: ,..,..,+1a
clean: ,..,..+
dirty: ,..,.,,+1a,,
clean: ,..,.,+,,
dirty: .$...,..,,,+1ga^:,
clean: ....,..,,+a,
dirty: .+1T,$.,.,...,,^].
clean: +,.,.,...,,.
dirty: ..,....,,+1aC
clean: ..,....,+C
dirty: ..,,,...-1G,,,
clean: ..,,,..-,,,
dirty: ,$,$,,-1t,
clean: ,,,-,
dirty: .+1T,..,,,,,
clean: +,..,,,,,
dirty: .+1T.,...^],
clean: +.,...,
dirty: ,,,.,,,,+1a,
clean: ,,,.,,,+,
dirty: ..,,+1g,.,,.^],
clean: ..,+,.,,.,
dirty: ,.+1T,,.,,,,,,,.,,,
clean: ,+,,.,,,,,,,.,,,
dirty: .+1T.^:.
clean: +..
```

* Function2: `get_relative_count`: called as `get_relative_count(<mpileup-file>)`.
    * Read in the `mpileup` file a row at a time
    * Save each row into an array of fields containing `chr`, `loc`, `ref`,
      `number`, `dirty bases`, `q1`, `q2`, `read positions`
    * Clean the `dirty bases` and store in `bases`
    * Split the `bases` into an array `tmp`, and
      `length_mutation` is this array's length
    * Split the `read positions` into an array `poss`
    * __If__ the coverage (number of reads) is equal to the `length_mutation` and
      within the min/max limits, do the following:
        * Split the `bases` into an array `base`
        * Split the `baseq` into an array `qualities`
        * For every index across the `base`:
            * Calculate the Phred quality using its baseq
            * If this passes the cutoff:
                * Collect the counts of ref-alt counts in a `result` hash with
                  two elements: `type` and `nt`:

* So essentially: go through the mpileup bases, and count the good quality forward
  strand mutations or reference calls. 'type' will be mutation type e.g. from T to G, or A to A (ref), or C to - (deletion).
  'nt' will be reference nucleotide total, for those mutations.

```
%result:
{
'type' => {
           'T_G' => 1,
           'G_G' => 57,
           'G_C' => 1,
           'G_T' => 2,
           'C_C' => 33,
           'A_A' => 123,
           'T_T' => 92,
           'G_A' => 1 },
'nt' => {
        'G' => 61,
        'A' => 123,
        'T' => 93,
        'C' => 33
        }
    };
```

* Close the file
* `mutations` are the different types
* Get all the different counts of types and store in the `final` hash:

```
%final:
{
  'T_G' => {
             'total' => 93,
             'total_type' => 1
           },
  'C_C' => {
             'total_type' => 33,
             'total' => 33
           },
  'A_A' => {
             'total' => 123,
             'total_type' => 123
           },
  'G_T' => {
             'total' => 61,
             'total_type' => 2
           },
  'G_G' => {
             'total_type' => 57,
             'total' => 61
           },
  'G_C' => {
             'total_type' => 1,
             'total' => 61
           },
  'T_T' => {
             'total' => 93,
             'total_type' => 92
           },
  'G_A' => {
             'total_type' => 1,
             'total' => 61
           }
};
```

### Main

* Get the `relative_count` from both mpileup files

This creates two hashes:

```
relative_count_R1:
$VAR1 = \{
            'T_-' => {
                       'total_type' => 1,
                       'total' => 7597
                     },
            'G_G' => {
                       'total_type' => 1784,
                       'total' => 1785
                     },
            'T_T' => {
                       'total_type' => 7596,
                       'total' => 7597
                     },
            'C_C' => {
                       'total_type' => 457,
                       'total' => 458
                     },
            'A_A' => {
                       'total_type' => 4134,
                       'total' => 4134
                     },
            'C_T' => {
                       'total' => 458,
                       'total_type' => 1
                     },
            'G_T' => {
                       'total_type' => 1,
                       'total' => 1785
                     }
          };

relative_count_R2:
$VAR1 = \{
            'A_G' => {
                       'total' => 1300,
                       'total_type' => 1
                     },
            'G_G' => {
                       'total' => 212,
                       'total_type' => 212
                     },
            'T_T' => {
                       'total_type' => 3710,
                       'total' => 3710
                     },
            'A_A' => {
                       'total_type' => 1299,
                       'total' => 1300
                     },
            'C_C' => {
                       'total_type' => 416,
                       'total' => 418
                     },
            'C_T' => {
                       'total' => 418,
                       'total_type' => 1
                     },
            'C_A' => {
                       'total' => 418,
                       'total_type' => 1
                     }
          };
```

* So up to here we've counted how many different mutations we have in
  both the R1 and R2 pileups.
* So each of the above hashes has two values for a given key-type: `total` and
  `total_type`. For each key-type in the R1 hash:
    * Get its reverse complement (`type_RC`)
    * Join `type` and `type_RC` (`couple`)
    * Get `total_type` of that type in R1 and R2 hashes
      (e.g. for `G_G` it will be 1784 for R1 and 416 for R2 (the reverse complement is taken for R2))
    * Get `total` similarly (e.g. for `G_G` it will be 1785/418 for R1/R2 RevComp)
    * If both R1/R2 have `total_type`, then:
        * Sum them up e.g. for `G_G`:
            * `total_type` = 1784 + 416 = 2200 
            * `total` = 1785 + 418 = 2203
        * Get the proportion:
            * `value` = `total_type` / `total`: 2200 / 2203 = 0.998
        * Put all that info into the `final` hash:

```
%final = {
  'A_A' => {
             'line' => '7844	A_A		1	A_A-T_T',
             'value' => '1'
           },
  'G_G' => {
             'value' => '0.998638220608261',
             'line' => '2200	G_G		0.998638220608261	C_C-G_G'
           },
  'C_T' => {
             'line' => '1	C_T		0.00218340611353712	C_T-G_A',
             'value' => '0.00218340611353712'
           },
  'T_T' => {
             'value' => '0.999775205125323',
             'line' => '8895	T_T		0.999775205125323	A_A-T_T'
           },
  'C_C' => {
             'line' => '669	C_C		0.998507462686567	C_C-G_G',
             'value' => '0.998507462686567'
           }
};
```

Now for each key-type in the `final` hash:

* Get its reverse complement (`type_RC`)
* If there are values for both `type` and `type_RC` in `final`:
    * Get the ratio of their values e.g. for `G_G` it will be:
    * `ratio1` = 0.998638220608261 / 0.998507462686567 = 1.000131


## Counting Mutations (more details)
We can compare the mutation counts in three different ways:

1. Total counts
2. Read position counts
3. Context counts

Each script takes as input two pileup files output by `samtools mpileup`,
corresponding to the first (flag = 64) and second (flag = 128) read of each pair.
There is one main function in each script which parses each line from the
mpileup files and counts each mutation type according to the items listed above.
The main output of each script is then generated by summarising those counts.

Total counts
-------------

### result hash

```
%result_R1     %result_R2
{
 'nt' => {
  'G' => 114,  'G' => 144,
  'C' => 111,  'A' => 263,
  'T' => 318,  'T' => 395,
  'A' => 283   'C' => 140
          },
 'type' => {
  'C_T' => 2,    'C_A' => 1,
  'C_C' => 108,  'G_G' => 144,
  'T_-' => 1,    'A_A' => 262,
  'A_A' => 283,  'T_T' => 394,
  'G_G' => 114,  'T_-' => 1,
  'C_A' => 1,    'C_T' => 2,
  'T_T' => 317   'A_T' => 1,
                 'C_C' => 137
            }
};

```

### final hash

* The `total_type` is the count from the `type` hash.
* The `total` is the count from the `nt` hash.

```
%final_R1 = {
'A_A' => { 'total' => 283, 'total_type' => 283 },
'C_C' => { 'total' => 111, 'total_type' => 108 },
'C_A' => { 'total' => 111, 'total_type' => 1},
'G_G' => { 'total' => 114, 'total_type' => 114 },
'C_T' => { 'total' => 111, 'total_type' => 2 },
'T_T' => { 'total' => 318, 'total_type' => 317 },
'T_-' => { 'total' => 318, 'total_type' => 1 }
        };

%final_R2 = {
'C_C' => { 'total' => 140, 'total_type' => 137},
'T_-' => { 'total' => 395, 'total_type' => 1 },
'C_T' => { 'total' => 140, 'total_type' => 2 },
'C_A' => { 'total' => 140, 'total_type' => 1 },
'A_A' => { 'total' => 263, 'total_type' => 262 },
'T_T' => { 'total' => 395, 'total_type' => 394 },
'G_G' => { 'total' => 144, 'total_type' => 144 },
'A_T' => { 'total' => 263, 'total_type' => 1 }
        };
```

### Output

* Take the two `%final` hashes output for R1 and R2.
* Iterate over the `%final_R1` type keys. 
* `type_rc` is the reverse complement of the type (e.g. `C_T` -> `G_A`)
* `couple`: `C_T-G_A`
* `total_type1`: the `total_type` from `%final_R1` for the given type
* `total_type2`: the `total_type` from `%final_R2` for the **complement** of the given type (if it exists)
* `total1`: the `total` from `%final_R1` for the given type
* `total2`: the `total` from `%final_R2` for the **complement** of the given type (if it exists)

* IF both `%final_R1` and `%final_R2` have `total_type` for the given type:
    * `total_type`: `total_type1` + `total_type2`
    * `total`: `total1` + `total2`
    * `value`: `total_type` / `total`
    * `line`: `total_type`, `type`, `id`, `value`, `couple`
    * Put `value` and `line` in the `final_final` hash for the given type
* Iterate over the `%final_final` hash
* IF there are values for the given type and its complement:
    * `ratio1` = `value` / `value_complement`
    * `line` = `line`
    * Print out `line`, `ratio1`

```
total_type  type    id  value   couple   ratio1
677         A_A     foo	0.998   A_A-T_T  1.001
252         C_C     foo	0.988   C_C-G_G  1.000
251         G_G     foo	0.988   C_C-G_G  0.999
579         T_T     foo	0.996   A_A-T_T  0.998
```

So essentially, for a given type `A_A` the `ratio1` is equal to:

* Numerator:   `R1[total_type(A_A)]` + `R2[total_type(T_T)]` / `R1[total(A)]` + `R2[total(T)]`
* Denominator: `R1[total_type(T_T)]` + `R2[total_type(A_A)]` / `R1[total(T)]` + `R2[total(A)]`


Read position counts
--------------------

### result hash

```
%result = {
          'nt' => {
                    '4' => {
                             'C' => 2,
                             'T' => 5
                           },
                    '113' => {
                               'A' => 1,
                               'T' => 2
                             },
                    '65' => {
                              'G' => 1,
                              'A' => 2,
                              'C' => 1,
                              'T' => 2
                            },

                    [...]

          'type' => {
                      '4' => {
                               'C_C' => 2,
                               'T_T' => 5
                             },
                      '113' => {
                                 'A_A' => 1,
                                 'T_T' => 2
                               },
                      '65' => {
                                'A_A' => 2,
                                'G_G' => 1,
                                'T_T' => 2,
                                'C_C' => 1
                              },
                    [...]

```

### final hash

* The `absolute` values are equal to `total_type`, which is from the `type` hash
  for the specific position.
* The `relative` values are `total_type / total`, where `total` is from the
  `nt` hash, for the specific position.

```
%final = {
          'C_T' => {
                     '74' => {
                               'absolute' => 1,
                               'relative' => '1'
                             },
                     '114' => {
                                'relative' => '0.5',
                                'absolute' => 1
                              }
                   },
          'A_A' => {
                     '13' => {
                               'relative' => '1',
                               'absolute' => 2
                             },
                     '59' => {
                               'absolute' => 4,
                               'relative' => '1'
                             },
                    [...]
```

### Output

* Iterate over `%RC`, where each key is a mutation type
* IF there are `relative` counts for the given type and position in `%final_R1`
  and `%final_R2`, then print out:
    * `id`, `type`, `read`, `relative`, `absolute`, `position`

Context counts
-------------

### result hash

```
          'type' => {
                      '89' => {
                                'A_A' => {
                                           'C_BASE_C' => 1,
                                           'T_BASE_A' => 1,
                                           'A_BASE_A' => 1
                                         },
                                'T_T' => {
                                           'G_BASE_T' => 1,
                                           'T_BASE_G' => 1,
                                           'A_BASE_T' => 1
                                         }
                              },
                      '61' => {
                                'A_A' => {
                                           'A_BASE_T' => 2,
                                           'T_BASE_A' => 1
                                         },
                                'T_T' => {
                                           'G_BASE_C' => 1,
                                           'A_BASE_A' => 1,
                                           'A_BASE_G' => 1
                                         },
                                'C_C' => {
                                           'T_BASE_C' => 1
                                         }
                              },
          'nt' => {
                    '89' => {
                              'A' => {
                                       'C_BASE_C' => 1,
                                       'T_BASE_A' => 1,
                                       'A_BASE_A' => 1
                                     },
                              'T' => {
                                       'A_BASE_T' => 1,
                                       'T_BASE_G' => 1,
                                       'G_BASE_T' => 1
                                     }
                            },
                    '61' => {
                              'A' => {
                                       'T_BASE_A' => 1,
                                       'A_BASE_T' => 2
                                     },
                              'T' => {
                                       'G_BASE_C' => 1,
                                       'A_BASE_A' => 1,
                                       'A_BASE_G' => 1
                                     },
                              'C' => {
                                       'T_BASE_C' => 1
                                     }
                            },
```
